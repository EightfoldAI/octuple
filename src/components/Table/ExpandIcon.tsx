import React from 'react';
import { Button, ButtonShape, ButtonVariant } from '../Button';
import { IconName } from '../Icon';
import { mergeClasses } from '../../shared/utilities';

import styles from './Styles/table.module.scss';

interface DefaultExpandIconProps<RecordType> {
  expandable: boolean;
  expanded: boolean;
  onExpand: (record: RecordType, e: React.MouseEvent<HTMLElement>) => void;
  record: RecordType;
  disabled?: boolean;
}

function renderExpandIcon(collapseText: string, expandText: string) {
  return function expandIcon<RecordType>({
    onExpand,
    record,
    expanded,
    expandable,
    disabled,
  }: DefaultExpandIconProps<RecordType>) {
    return (
      <Button
        ariaLabel={expanded ? collapseText : expandText}
        classNames={mergeClasses([
          styles.tableRowExpandIcon,
          { [styles.tableRowExpandIconVisuallyHidden]: !expandable },
          { [styles.tableExpandedRow]: expandable && expanded },
        ])}
        disabled={disabled}
        htmlType="button"
        iconProps={{
          path: expanded ? IconName.mdiChevronUp : IconName.mdiChevronDown,
        }}
        onClick={(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
          if (disabled) {
            return;
          }
          onExpand(record, e!);
          e?.stopPropagation();
        }}
        shape={ButtonShape.Round}
        variant={ButtonVariant.Neutral}
      />
    );
  };
}

export default renderExpandIcon;
