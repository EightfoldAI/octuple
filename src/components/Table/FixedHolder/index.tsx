import React, {
    forwardRef,
    useCallback,
    useEffect,
    useContext,
    useMemo,
    useRef,
} from 'react';
import { mergeClasses } from '../../../shared/utilities';
import { fillRef } from '../../../shared/ref';
import ColGroup from '../ColGroup';
import type { ColumnsType, ColumnType } from '../Table.types';
import { FixedHeaderProps } from './FixedHolder.types';
import TableContext from '../Context/TableContext';

import styles from './Table.module.scss';

function useColumnWidth(colWidths: readonly number[], columCount: number) {
    return useMemo(() => {
        const cloneColumns: number[] = [];
        for (let i = 0; i < columCount; i += 1) {
            const val = colWidths[i];
            if (val !== undefined) {
                cloneColumns[i] = val;
            } else {
                return null;
            }
        }
        return cloneColumns;
    }, [colWidths.join('_'), columCount]);
}

const FixedHolder = forwardRef<HTMLDivElement, FixedHeaderProps<unknown>>(
    (
        {
            classNames,
            noData,
            columns,
            flattenColumns,
            colWidths,
            columCount,
            stickyOffsets,
            direction,
            fixHeader,
            stickyTopOffset,
            stickyBottomOffset,
            stickyClassName,
            onScroll,
            maxContentScroll,
            children,
            ...props
        },
        ref
    ) => {
        const { scrollbarSize, isSticky } = useContext(TableContext);

        const combinationScrollBarSize =
            isSticky && !fixHeader ? 0 : scrollbarSize;

        // Pass wheel to scroll event
        const scrollRef = useRef<HTMLDivElement>(null);

        const setScrollRef = useCallback((element: HTMLElement) => {
            fillRef(ref, element);
            fillRef(scrollRef, element);
        }, []);

        useEffect(() => {
            function onWheel(e: WheelEvent) {
                const { currentTarget, deltaX } =
                    e as unknown as React.WheelEvent<HTMLDivElement>;
                if (deltaX) {
                    onScroll({
                        currentTarget,
                        scrollLeft: currentTarget.scrollLeft + deltaX,
                    });
                    e.preventDefault();
                }
            }
            scrollRef.current?.addEventListener('wheel', onWheel);

            return () => {
                scrollRef.current?.removeEventListener('wheel', onWheel);
            };
        }, []);

        // Check if all flattenColumns has width
        const allFlattenColumnsWithWidth = useMemo(
            () => flattenColumns.every((column) => column.width >= 0),
            [flattenColumns]
        );

        // Add scrollbar column
        const lastColumn = flattenColumns[flattenColumns.length - 1];
        const ScrollBarColumn: ColumnType<unknown> & { scrollbar: true } = {
            fixed: lastColumn ? lastColumn.fixed : null,
            scrollbar: true,
            onHeaderCell: () => ({
                className: styles.tableCellScrollbar,
            }),
        };

        const columnsWithScrollbar = useMemo<ColumnsType<unknown>>(
            () =>
                combinationScrollBarSize
                    ? [...columns, ScrollBarColumn]
                    : columns,
            [combinationScrollBarSize, columns]
        );

        const flattenColumnsWithScrollbar = useMemo(
            () =>
                combinationScrollBarSize
                    ? [...flattenColumns, ScrollBarColumn]
                    : flattenColumns,
            [combinationScrollBarSize, flattenColumns]
        );

        // Calculate the sticky offsets
        const headerStickyOffsets = useMemo(() => {
            const { right, left } = stickyOffsets;
            return {
                ...stickyOffsets,
                left:
                    direction === 'rtl'
                        ? [
                              ...left.map(
                                  (width) => width + combinationScrollBarSize
                              ),
                              0,
                          ]
                        : left,
                right:
                    direction === 'rtl'
                        ? right
                        : [
                              ...right.map(
                                  (width) => width + combinationScrollBarSize
                              ),
                              0,
                          ],
                isSticky,
            };
        }, [combinationScrollBarSize, stickyOffsets, isSticky]);

        const mergedColumnWidth = useColumnWidth(colWidths, columCount);

        return (
            <div
                style={{
                    overflow: 'hidden',
                    ...(isSticky
                        ? { top: stickyTopOffset, bottom: stickyBottomOffset }
                        : {}),
                }}
                ref={setScrollRef}
                className={mergeClasses([
                    classNames,
                    { [stickyClassName]: !!stickyClassName },
                ])}
            >
                <table
                    style={{
                        tableLayout: 'fixed',
                        visibility:
                            noData || mergedColumnWidth ? null : 'hidden',
                    }}
                >
                    {(!noData ||
                        !maxContentScroll ||
                        allFlattenColumnsWithWidth) && (
                        <ColGroup
                            colWidths={
                                mergedColumnWidth
                                    ? [
                                          ...mergedColumnWidth,
                                          combinationScrollBarSize,
                                      ]
                                    : []
                            }
                            columCount={columCount + 1}
                            columns={flattenColumnsWithScrollbar}
                        />
                    )}
                    {children({
                        ...props,
                        stickyOffsets: headerStickyOffsets,
                        columns: columnsWithScrollbar,
                        flattenColumns: flattenColumnsWithScrollbar,
                    })}
                </table>
            </div>
        );
    }
);

FixedHolder.displayName = 'FixedHolder';

export default FixedHolder;
