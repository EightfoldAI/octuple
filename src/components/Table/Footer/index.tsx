import React, { useMemo } from 'react';
import Summary from './Summary';
import SummaryContext from './SummaryContext';
import { FooterProps } from './Footer.types';

import styles from '../table.module.scss';

function Footer<RecordType>({
    children,
    stickyOffsets,
    flattenColumns,
}: FooterProps<RecordType>) {
    const lastColumnIndex = flattenColumns.length - 1;
    const scrollColumn = flattenColumns[lastColumnIndex];

    const summaryContext = useMemo(
        () => ({
            stickyOffsets,
            flattenColumns,
            scrollColumnIndex: scrollColumn?.scrollbar ? lastColumnIndex : null,
        }),
        [scrollColumn, flattenColumns, lastColumnIndex, stickyOffsets]
    );

    return (
        <SummaryContext.Provider value={summaryContext}>
            <tfoot className={styles.tableSummary}>{children}</tfoot>
        </SummaryContext.Provider>
    );
}

export default Footer;

export const FooterComponents = Summary;
