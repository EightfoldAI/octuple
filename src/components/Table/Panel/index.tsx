import React from 'react';
import { TitleProps } from './Panel.types';

function Panel({ classNames, children }: TitleProps) {
    return <div className={classNames}>{children}</div>;
}

export default Panel;
