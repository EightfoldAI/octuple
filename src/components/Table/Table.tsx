import React, {
    forwardRef,
    ReactNode,
    useCallback,
    useContext,
    useMemo,
    useRef,
} from 'react';
import type { Breakpoint } from '../../shared/utilities';
import { mergeClasses, omit, scrollTo } from '../../shared/utilities';
import OcTable, { Summary } from './Internal';
import type { OcTableProps } from './Internal/OcTable.types';
import { convertChildrenToColumns } from './Internal/Hooks/useColumns';
import { Spinner, SpinnerSize } from '../Spinner';
import { Pagination } from '../Pagination';
import usePagination, {
    DEFAULT_PAGE_SIZE,
    getPaginationParam,
} from './Hooks/usePagination';
import useLazyKVMap from './Hooks/useLazyKVMap';
import type {
    ChangeEventInfo,
    ColumnGroupType,
    ColumnType,
    ExpandableConfig,
    ExpandType,
    FilterValue,
    GetRowKey,
    SorterResult,
    TableAction,
    TableProps,
} from './Table.types';
import {
    EMPTY_LIST,
    ColumnsType,
    TablePaginationConfig,
    TableSize,
} from './Table.types';
import useSelection, {
    SELECTION_ALL,
    SELECTION_COLUMN,
    SELECTION_INVERT,
    SELECTION_NONE,
} from './Hooks/useSelection';
import type { SortState } from './Hooks/useSorter';
import useSorter, { getSortData } from './Hooks/useSorter';
import type { FilterState } from './Hooks/useFilter';
import useFilter, { getFilterData } from './Hooks/useFilter';
import useTitleColumns from './Hooks/useTitleColumns';
import renderExpandIcon from './ExpandIcon';
import { Size, SizeContext } from '../ConfigProvider';
import Column from './Internal/Column';
import ColumnGroup from './Internal/ColumnGroup';
import { useBreakpoint } from '../../hooks/useBreakpoint';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { Empty, EmptyMode } from '../Empty/index';

import styles from './Styles/table.module.scss';

export { ColumnsType, TablePaginationConfig };

function InternalTable<RecordType extends object = any>(
    props: TableProps<RecordType>,
    ref: React.MutableRefObject<HTMLDivElement>
) {
    const {
        alternateRowColor = true,
        bordered,
        cancelSortText,
        cellBordered = false,
        children,
        classNames,
        collapseText,
        columns,
        dataSource,
        emptyText,
        emptyTextDetails,
        expandableConfig,
        expandText,
        filterCheckallText,
        filterConfirmText,
        filterEmptyText,
        filterResetText,
        filterSearchPlaceholderText,
        getPopupContainer,
        headerBordered = false,
        headerBottomBordered = false,
        indentSize,
        innerBordered = false,
        loading,
        onChange,
        outerBordered = false,
        pagination,
        rowBordered = false,
        rowClassName,
        rowKey,
        rowSelection,
        scroll,
        selectInvertText,
        selectionAllText,
        selectNoneText,
        showSorterTooltip = true,
        size = TableSize.Medium,
        sortDirections,
        style,
        triggerAscText,
        triggerDescText,
    } = props;

    const baseColumns: ColumnsType<RecordType> = useMemo(
        () =>
            columns ||
            (convertChildrenToColumns(children) as ColumnsType<RecordType>),
        [columns, children]
    );
    const needResponsive: boolean = useMemo(
        () => baseColumns.some((col: ColumnType<RecordType>) => col.responsive),
        [baseColumns]
    );

    const screens: Partial<Record<Breakpoint, boolean>> =
        useBreakpoint(needResponsive);

    const mergedColumns: (
        | ColumnGroupType<RecordType>
        | ColumnType<RecordType>
    )[] = useMemo(() => {
        const matched = new Set(
            Object.keys(screens).filter((m: Breakpoint) => screens[m])
        );

        return baseColumns.filter(
            (c) =>
                !c.responsive ||
                c.responsive.some((r: Breakpoint) => matched.has(r))
        );
    }, [baseColumns, screens]);

    const tableProps = omit(props, [
        'classNames',
        'style',
        'columns',
    ]) as TableProps<RecordType>;

    const contextuallySized: Size = useContext(SizeContext);
    const mergedSize = contextuallySized || size;

    const htmlDir: string = useCanvasDirection();
    const rawData: readonly RecordType[] = dataSource || EMPTY_LIST;

    const mergedExpandableConfig: ExpandableConfig<RecordType> = {
        ...expandableConfig,
    };

    const { childrenColumnName = 'children' } = mergedExpandableConfig;

    const expandType: ExpandType = useMemo<ExpandType>(() => {
        if (rawData.some((item) => (item as any)?.[childrenColumnName])) {
            return 'nest';
        }

        if (expandableConfig?.expandedRowRender) {
            return 'row';
        }

        return null;
    }, [rawData]);

    const internalRefs = {
        body: useRef<HTMLDivElement>(),
    };

    // ============================ RowKey ============================
    const getRowKey = useMemo<GetRowKey<RecordType>>(() => {
        if (typeof rowKey === 'function') {
            return rowKey;
        }

        return (record: RecordType) => (record as any)?.[rowKey as string];
    }, [rowKey]);

    const [getRecordByKey] = useLazyKVMap(
        rawData,
        childrenColumnName,
        getRowKey
    );

    // ============================ Events =============================
    const changeEventInfo: Partial<ChangeEventInfo<RecordType>> = {};

    const triggerOnChange = (
        info: Partial<ChangeEventInfo<RecordType>>,
        action: TableAction,
        reset: boolean = false
    ) => {
        const changeInfo = {
            ...changeEventInfo,
            ...info,
        };

        if (reset) {
            changeEventInfo.resetPagination!();

            // Reset event param
            if (changeInfo.pagination!.currentPage) {
                changeInfo.pagination!.currentPage = 1;
            }

            // Trigger pagination events
            if (pagination) {
                pagination.onCurrentChange?.(
                    changeInfo.pagination!.currentPage!
                );

                const pages: number[] = changeInfo.pagination!.pageSizes!;

                if (pages) {
                    console.log('triggerOnChange pages: ' + pages);
                    for (let i: number = 0; i < pages.length; ++i) {
                        pagination.onSizeChange?.(
                            changeInfo.pagination!.pageSizes[i]
                        );
                    }
                } else {
                    pagination.onSizeChange?.(changeInfo.pagination!.pageSize!);
                }
            }
        }

        if (
            scroll &&
            scroll.scrollToFirstRowOnChange !== false &&
            internalRefs.body.current
        ) {
            scrollTo(0, {
                getContainer: () => internalRefs.body.current!,
            });
        }

        onChange?.(
            changeInfo.pagination!,
            changeInfo.filters!,
            changeInfo.sorter!,
            {
                currentDataSource: getFilterData(
                    getSortData(
                        rawData,
                        changeInfo.sorterStates!,
                        childrenColumnName
                    ),
                    changeInfo.filterStates!
                ),
                action,
            }
        );
    };

    // ============================ Sorter =============================
    const onSorterChange = (
        sorter: SorterResult<RecordType> | SorterResult<RecordType>[],
        sorterStates: SortState<RecordType>[]
    ) => {
        triggerOnChange(
            {
                sorter,
                sorterStates,
            },
            'sort',
            false
        );
    };
    const [transformSorterColumns, sortStates, sorterTitleProps, getSorters] =
        useSorter<RecordType>({
            cancelSortText,
            mergedColumns,
            onSorterChange,
            sortDirections: sortDirections || ['ascend', 'descend'],
            triggerAscText,
            triggerDescText,
            showSorterTooltip,
        });
    const sortedData = useMemo(
        () => getSortData(rawData, sortStates, childrenColumnName),
        [rawData, sortStates]
    );

    changeEventInfo.sorter = getSorters();
    changeEventInfo.sorterStates = sortStates;

    // ============================ Filter ============================
    const onFilterChange = (
        filters: Record<string, FilterValue>,
        filterStates: FilterState<RecordType>[]
    ) => {
        triggerOnChange(
            {
                filters,
                filterStates,
            },
            'filter',
            true
        );
    };

    const [transformFilterColumns, filterStates, getFilters] =
        useFilter<RecordType>({
            mergedColumns,
            filterConfirmText,
            filterResetText,
            filterEmptyText,
            filterCheckallText,
            filterSearchPlaceholderText,
            onFilterChange,
            getPopupContainer,
        });
    const mergedData = getFilterData(sortedData, filterStates);

    changeEventInfo.filters = getFilters();
    changeEventInfo.filterStates = filterStates;

    // ============================ Column ============================
    const columnTitleProps = useMemo(
        () => ({
            ...sorterTitleProps,
        }),
        [sorterTitleProps]
    );
    const [transformTitleColumns] = useTitleColumns(columnTitleProps);

    // ========================== Pagination ==========================
    const onPaginationChange = (
        currentPage: number,
        pageSize: number,
        pageSizes: number[]
    ) => {
        console.log(
            'onPaginationChange pageSize: ' +
                pageSize +
                ' ' +
                'onPaginationChange pageSizes: ' +
                pageSizes
        );
        triggerOnChange(
            {
                pagination: {
                    ...changeEventInfo.pagination,
                    currentPage,
                    pageSize,
                    pageSizes,
                },
            },
            'paginate'
        );
    };

    const [mergedPagination, resetPagination] = usePagination(
        mergedData.length,
        pagination,
        onPaginationChange
    );

    changeEventInfo.pagination =
        pagination === false
            ? {}
            : getPaginationParam(pagination, mergedPagination);

    changeEventInfo.resetPagination = resetPagination;

    // ============================= Data =============================
    const pageData = useMemo<RecordType[]>(() => {
        if (
            pagination === false ||
            (!mergedPagination.pageSize && !mergedPagination.pageSizes)
        ) {
            return mergedData;
        }

        const {
            currentPage = 1,
            total,
            pageSize = DEFAULT_PAGE_SIZE, // Need to figure out how to update this value
            pageSizes,
        } = mergedPagination;

        // Dynamic table data
        if (pageSizes) {
            console.log('Dynamic table data pageSizes: ' + pageSizes);
            for (let i: number = 0; i < pageSizes.length; ++i) {
                if (pageSize === pageSizes[i]) {
                    console.log(
                        'Dynamic table data pageSizes[' +
                            i +
                            '] value: ' +
                            pageSizes[i]
                    );
                    if (mergedData.length < total!) {
                        if (mergedData.length > pageSizes[i]) {
                            return mergedData.slice(
                                (currentPage - 1) * pageSizes[i],
                                currentPage * pageSizes[i]
                            );
                        }
                        return mergedData;
                    }

                    console.log(
                        mergedData.slice(
                            (currentPage - 1) * pageSizes[i],
                            currentPage * pageSizes[i]
                        )
                    );

                    return mergedData.slice(
                        (currentPage - 1) * pageSizes[i],
                        currentPage * pageSizes[i]
                    );
                }
            }
        } else {
            console.log('Dynamic table data pageSize: ' + pageSize);
            if (mergedData.length < total!) {
                if (mergedData.length > pageSize) {
                    return mergedData.slice(
                        (currentPage - 1) * pageSize,
                        currentPage * pageSize
                    );
                }
                return mergedData;
            }

            return mergedData.slice(
                (currentPage - 1) * pageSize,
                currentPage * pageSize
            );
        }
        return null;
    }, [
        !!pagination,
        mergedData,
        mergedPagination?.currentPage,
        mergedPagination?.pageSize,
        mergedPagination?.pageSizes,
        mergedPagination?.total,
    ]);

    // ========================== Selections ==========================
    const [transformSelectionColumns, selectedKeySet] =
        useSelection<RecordType>(rowSelection, {
            data: mergedData,
            emptyText,
            emptyTextDetails,
            pageData,
            getRowKey,
            getRecordByKey,
            expandType,
            childrenColumnName,
            getPopupContainer,
            selectionAllText,
            selectInvertText,
            selectNoneText,
        });

    const internalRowClassName = (
        record: RecordType,
        index: number,
        indent: number
    ) => {
        let mergedRowClassName;
        if (typeof rowClassName === 'function') {
            mergedRowClassName = mergeClasses([
                rowClassName(record, index, indent),
            ]);
        } else {
            mergedRowClassName = mergeClasses([rowClassName]);
        }

        return mergeClasses(
            {
                [styles.tableRowSelected]: selectedKeySet.has(
                    getRowKey(record, index)
                ),
            },
            mergedRowClassName
        );
    };

    // ========================== Expandable ==========================

    mergedExpandableConfig.expandIcon = renderExpandIcon(
        expandText,
        collapseText
    );

    // Indent size
    if (typeof mergedExpandableConfig.indentSize !== 'number') {
        mergedExpandableConfig.indentSize =
            typeof indentSize === 'number' ? indentSize : 15;
    }

    // ============================ Render ============================
    const transformColumns = useCallback(
        (innerColumns: ColumnsType<RecordType>): ColumnsType<RecordType> =>
            transformTitleColumns(
                transformSelectionColumns(
                    transformFilterColumns(transformSorterColumns(innerColumns))
                )
            ),
        [
            transformSorterColumns,
            transformFilterColumns,
            transformSelectionColumns,
        ]
    );

    let topPaginationNode: React.ReactNode;
    let bottomPaginationNode: React.ReactNode;

    if (pagination !== false && mergedPagination?.total) {
        let paginationSize: TablePaginationConfig['pageSize'];
        let paginationSizes: TablePaginationConfig['pageSizes'];

        if (mergedPagination.pageSize) {
            console.log('Table pageSize: ' + mergedPagination.pageSize);
            paginationSize = mergedPagination.pageSize;
        } else {
            paginationSize = undefined;
        }
        if (mergedPagination.pageSizes) {
            console.log('Table pageSizes: ' + mergedPagination.pageSizes);
            paginationSizes = mergedPagination.pageSizes;
        } else {
            paginationSizes = undefined;
        }

        const renderPagination = (position: string) => (
            <Pagination
                {...mergedPagination}
                classNames={mergeClasses([
                    styles.tablePagination,
                    {
                        [styles.tablePaginationLeft]:
                            position.indexOf('left') !== -1,
                    },
                    {
                        [styles.tablePaginationCenter]:
                            position.indexOf('center') !== -1,
                    },
                    {
                        [styles.tablePaginationRight]:
                            position.indexOf('right') !== -1,
                    },
                    mergedPagination.className,
                ])}
                pageSize={paginationSize}
                pageSizes={paginationSizes}
                total={mergedPagination?.total}
            />
        );
        const defaultPosition = htmlDir === 'rtl' ? 'left' : 'right';
        const { position } = mergedPagination;
        if (position !== null && Array.isArray(position)) {
            const topPos = position.find((p) => p.indexOf('top') !== -1);
            const bottomPos = position.find((p) => p.indexOf('bottom') !== -1);
            const isDisable = position.every((p) => `${p}` === 'none');
            if (!topPos && !bottomPos && !isDisable) {
                bottomPaginationNode = renderPagination(defaultPosition);
            }
            if (topPos) {
                topPaginationNode = renderPagination(
                    topPos!.toLowerCase().replace('top', '')
                );
            }
            if (bottomPos) {
                bottomPaginationNode = renderPagination(
                    bottomPos!.toLowerCase().replace('bottom', '')
                );
            }
        } else {
            bottomPaginationNode = renderPagination(defaultPosition);
        }
    }

    if (typeof loading === 'boolean') {
        return <Spinner size={SpinnerSize.Large} />;
    } else if (typeof loading === 'object') {
        return <Spinner size={SpinnerSize.Large} />;
    }

    const renderEmpty = (
        emptyText: ReactNode,
        emptyTextDetails?: string
    ): ReactNode => {
        if (typeof emptyText === 'string') {
            if (emptyTextDetails) {
                return (
                    <Empty
                        title={emptyText}
                        description={emptyTextDetails}
                        mode={EmptyMode.data}
                    />
                );
            }
            return <Empty title={emptyText} mode={EmptyMode.data} />;
        } else {
            return emptyText;
        }
    };

    const wrapperClassNames: string = mergeClasses([
        styles.tableWrapper,
        { [styles.tableWrapperRtl]: htmlDir === 'rtl' },
        classNames,
    ]);
    return (
        <div ref={ref} className={wrapperClassNames} style={style}>
            {topPaginationNode}
            <OcTable<RecordType>
                {...tableProps}
                columns={mergedColumns as OcTableProps<RecordType>['columns']}
                direction={htmlDir}
                expandableConfig={mergedExpandableConfig}
                classNames={mergeClasses([
                    styles.table,
                    { [styles.tableRtl]: htmlDir === 'rtl' },
                    { [styles.tableLarge]: mergedSize === TableSize.Large },
                    { [styles.tableMedium]: mergedSize === TableSize.Medium },
                    { [styles.tableSmall]: mergedSize === TableSize.Small },
                    { [styles.tableAlternate]: alternateRowColor },
                    { [styles.tableBordered]: bordered },
                    {
                        [styles.tableCellBordered]:
                            !bordered &&
                            !rowBordered &&
                            !innerBordered &&
                            cellBordered,
                    },
                    {
                        [styles.tableHeaderBordered]:
                            !bordered &&
                            !rowBordered &&
                            !innerBordered &&
                            headerBordered,
                    },
                    {
                        [styles.tableHeaderBottomBordered]:
                            !bordered &&
                            !rowBordered &&
                            !innerBordered &&
                            !headerBordered &&
                            headerBottomBordered,
                    },
                    {
                        [styles.tableInnerBordered]:
                            !bordered && !rowBordered && innerBordered,
                    },
                    { [styles.tableOuterBordered]: !bordered && outerBordered },
                    {
                        [styles.tableRowBordered]:
                            !bordered && !innerBordered && rowBordered,
                    },
                    { [styles.tableEmpty]: rawData.length === 0 },
                ])}
                data={pageData}
                rowKey={getRowKey}
                rowClassName={internalRowClassName}
                emptyText={renderEmpty(emptyText, emptyTextDetails)}
                transformColumns={
                    transformColumns as OcTableProps<RecordType>['transformColumns']
                }
            />
            {bottomPaginationNode}
        </div>
    );
}

const ForwardTable = forwardRef(InternalTable) as <
    RecordType extends object = any
>(
    props: React.PropsWithChildren<TableProps<RecordType>> & {
        ref?: React.Ref<HTMLDivElement>;
    }
) => React.ReactElement;

type InternalTableType = typeof ForwardTable;

interface TableInterface extends InternalTableType {
    defaultProps?: Partial<TableProps<any>>;
    SELECTION_COLUMN: typeof SELECTION_COLUMN;
    EXPAND_COLUMN: typeof OcTable.EXPAND_COLUMN;
    SELECTION_ALL: 'SELECT_ALL';
    SELECTION_INVERT: 'SELECT_INVERT';
    SELECTION_NONE: 'SELECT_NONE';
    Column: typeof Column;
    ColumnGroup: typeof ColumnGroup;
    Summary: typeof Summary;
}

const Table = ForwardTable as TableInterface;

Table.defaultProps = {
    filterConfirmText: 'OK',
    filterResetText: 'Reset',
    filterEmptyText: 'No filters',
    filterCheckallText: 'Select all items',
    filterSearchPlaceholderText: 'Search in filters',
    emptyText: 'Short Message Here',
    emptyTextDetails:
        'More detail on how might the user be able to get around this',
    selectInvertText: 'Invert current page',
    selectNoneText: 'Clear all data',
    selectionAllText: 'Select all data',
    expandText: 'Expand row',
    collapseText: 'Collapse row',
    triggerDescText: 'Click to sort descending',
    triggerAscText: 'Click to sort ascending',
    cancelSortText: 'Click to cancel sorting',
    rowKey: 'key',
};

Table.SELECTION_COLUMN = SELECTION_COLUMN;
Table.EXPAND_COLUMN = OcTable.EXPAND_COLUMN;
Table.SELECTION_ALL = SELECTION_ALL;
Table.SELECTION_INVERT = SELECTION_INVERT;
Table.SELECTION_NONE = SELECTION_NONE;
Table.Column = Column;
Table.ColumnGroup = ColumnGroup;
Table.Summary = Summary;

export default Table;
