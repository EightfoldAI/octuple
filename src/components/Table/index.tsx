import Table, { TablePaginationConfig } from './Table';

export { ColumnProps } from './Internal/Column';
export {
    ColumnsType,
    ColumnType,
    ColumnGroupType,
    ExpandableConfig,
    FilterConfirmProps,
    FilterValue,
    SorterResult,
    TableProps,
    TableRowSelection,
    TableSize,
} from './Table.types';
export { TablePaginationConfig };

export default Table;
