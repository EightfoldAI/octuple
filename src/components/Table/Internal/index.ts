import OcTable from './OcTable';
import { FooterComponents as Summary } from './Footer';
import Column from './Column';
import ColumnGroup from './ColumnGroup';
import { INTERNAL_COL_DEFINE } from './constant';

export { Summary, Column, ColumnGroup, INTERNAL_COL_DEFINE };

export default OcTable;
