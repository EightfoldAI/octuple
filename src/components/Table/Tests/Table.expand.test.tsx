/* eslint-disable react/no-multi-comp */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Table from '../index';
import { Button, ButtonShape, ButtonVariant } from '../../Button';
import { IconName } from '../../Icon';

Enzyme.configure({ adapter: new Adapter() });

const columns = [
  {
    title: 'Name',
    key: 'name',
    dataIndex: 'name',
  },
];

const John = {
  key: '1',
  firstName: 'John',
  lastName: 'Brown',
  age: 32,
};

const Jim = {
  key: '2',
  firstName: 'Jim',
  lastName: 'Green',
  age: 42,
};

const data = [
  {
    ...John,

    children: [
      {
        ...Jim,
      },
    ],
  },
];

describe('Table.expand', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('click to expand', () => {
    const wrapper = mount(<Table columns={columns} dataSource={data} />);
    wrapper.find('.table-row-expand-icon').last().simulate('click');
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('row indent padding should be 0px when indentSize defined as 0', () => {
    const wrapper = mount(
      <Table
        indentSize={0}
        columns={columns}
        dataSource={data}
        expandableConfig={{
          expandIcon: ({
            disabled,
            expandable,
            expanded,
            onExpand,
            record,
          }: {
            disabled: boolean;
            expandable: boolean;
            expanded: boolean;
            onExpand: (record: any, e: React.MouseEvent<HTMLElement>) => void;
            record: any;
          }) => {
            return (
              <Button
                ariaLabel={expanded ? 'Collapse row' : 'Expand row'}
                classNames="expand-icon"
                disabled={disabled}
                htmlType="button"
                iconProps={{
                  path: expanded
                    ? IconName.mdiChevronUp
                    : IconName.mdiChevronDown,
                }}
                onClick={(
                  e: React.MouseEvent<HTMLButtonElement, MouseEvent>
                ) => {
                  if (disabled) {
                    return;
                  }
                  onExpand(record, e!);
                  e?.stopPropagation();
                }}
                shape={ButtonShape.Round}
                style={{
                  display: expandable ? 'flex' : 'none',
                }}
                variant={ButtonVariant.Neutral}
              />
            );
          },
        }}
      />
    );
    const button = wrapper.find('.expand-icon').at(0);
    button.simulate('click');
    expect(wrapper.find('.indent-level-1').at(0).prop('style')).toHaveProperty(
      'paddingLeft',
      '0px'
    );
  });
});
