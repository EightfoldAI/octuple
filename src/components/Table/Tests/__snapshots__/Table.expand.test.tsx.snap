version https://git-lfs.github.com/spec/v1
oid sha256:0b5958c450632c524efec465d2050e94d61b2a639c02b475b04c2ec9005bc4a2
size 229560
